include/revng/Pipes/PipelineManager.h

```diff
@@ -0,0 +1,101 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <string>
+#include <vector>
+
+#include "llvm/ADT/StringRef.h"
+
+#include "revng/Pipeline/Context.h"
+#include "revng/Pipeline/Runner.h"
+#include "revng/Pipeline/SavableObject.h"
+#include "revng/Pipes/LLVMContextWrapper.h"
+#include "revng/Pipes/ModelGlobal.h"
+
+namespace revng::pipes {
+
+/// This is a "God Object" that can be used when there will be exactly one
+/// Pipeline spawned by this process.
+///
+/// It will care of all configurations including llvm ones and provide a runner
+/// in a usable state.
+class PipelineManager {
+private:
+  explicit PipelineManager() = default;
+
+  std::string ExecutionDirectory;
+  std::unique_ptr<LLVMContextWrapper> Context;
+  std::unique_ptr<ModelGlobal> ModelWrapper;
+  std::unique_ptr<pipeline::Context> PipelineContext;
+  std::unique_ptr<pipeline::Loader> Loader;
+  std::unique_ptr<pipeline::Runner> Pipeline;
+  pipeline::Runner::State State;
+  std::map<const pipeline::ContainerSet::value_type *,
+           const pipeline::TargetsList *>
+    LookUpMap;
+
+  static llvm::Expected<PipelineManager>
+  createContexts(llvm::ArrayRef<std::string> EnablingFlags,
+                 llvm::StringRef ExecutionDirectory);
+
+public:
+  PipelineManager(PipelineManager &&Other) = default;
+  PipelineManager &operator=(PipelineManager &&Other) = default;
+  PipelineManager &operator=(const PipelineManager &Other) = delete;
+  PipelineManager(const PipelineManager &Other) = delete;
+  ~PipelineManager() = default;
+
+  const pipeline::Kind *getKind(llvm::StringRef Name) const {
+    return Pipeline->getKindsRegistry().find(Name);
+  }
+
+  static llvm::Expected<PipelineManager>
+  create(llvm::ArrayRef<std::string> PipelinePath,
+         llvm::ArrayRef<std::string> EnablingFlags,
+         llvm::StringRef ExecutionDirectory);
+
+  static llvm::Expected<PipelineManager>
+  createFromMemory(llvm::ArrayRef<std::string> InMemoryPipeline,
+                   llvm::ArrayRef<std::string> EnablingFlags,
+                   llvm::StringRef ExecutionDirectory);
+
+  llvm::Error overrideContainer(llvm::StringRef PipelineFileMapping);
+  llvm::Error overrideContainer(pipeline::PipelineFileMapping Mapping);
+  llvm::Error overrideModel(llvm::StringRef Path);
+
+  llvm::Error storeOverrides(llvm::ArrayRef<std::string> StoresOverrides);
+  llvm::Error
+  storeOverride(const pipeline::PipelineFileMapping &StoresOverride);
+  llvm::Error storeToDisk();
+
+  const pipeline::Context &context() const { return *PipelineContext; }
+
+  void recalculateAllPossibleTargets();
+  void recalculateCurrentState();
+  void recalculateCache();
+  void getAllPossibleTargets(pipeline::Runner::State &State) const;
+  void getCurrentState(pipeline::Runner::State &State) const;
+  const pipeline::Runner::State &getLastPipelineState() const { return State; }
+  llvm::Error produceAllPossibleTargets();
```

Comment:
What's the use case for this? Is it like `make all`?

Reply:
it's used for debug purposes to require the production of everything that can be produced and detect crashes in a reliable way

Reply:
OK, document this. The public API of this class otherwise is quite confusing.